### DISCUSSION TITLE
How to return either success or errors?

### DISCUSSION THREAD

**Discussion by tobymurray** [rwf2/Rocket#2210] [Questions]
I'm just not getting this - it seems like I'm missing something obvious. I want to return either a successful `rocket_dyn_templates::Template` or an error (e.g. 401, 403, 500), and the slight twist is I want to add a custom header as well.

First:

> Responders may fail instead of generating a response by returning an Err with a status code

What does this mean in practice? What's the response signature for this method?

> a default error catcher will be used

If I declare e.g. `Result<Template, String>` and return `Err("This is an error".to_string()`, I get an HTTP 200 response with this content. The docs read as though I should expect Rocket to be _doing_ something with the `Err` that I'm returning (maybe a 500 or some other default?). Does the default error handler catch errors and turn them into HTTP 200s?

Elsewhere the docs suggest:

> If you're returning two different responses, use a Result<T, E> or an Either<A, B>.
> 
> If you need to return more than two kinds, derive a custom Responder enum:
> rocket_dyn_templates::Template

```
use rocket::fs::NamedFile;
use rocket::http::ContentType;

#[derive(Responder)]
enum Error<'r, T> {
    #[response(status = 400)]
    Unauthorized(T),
    #[response(status = 404)]
    NotFound(NamedFile),
    #[response(status = 500)]
    A(&'r str, ContentType),
}
```

But this is _just_ the error. Is this example showing how to implement an API (`Responder`) that exclusively returns errors? If so... that's a very confusing piece of documentation, as it seems unlikely anyone would practically use that.

I've seen [the error handling example](https://github.com/SergioBenitez/Rocket/tree/v0.5-rc/examples/error-handling/src), https://github.com/SergioBenitez/Rocket/issues/71, https://github.com/SergioBenitez/Rocket/issues/72, https://github.com/SergioBenitez/Rocket/discussions/2155, and https://github.com/SergioBenitez/Rocket/discussions/2172 and for the life of me I can't put the pieces together.

**Comment by tobymurray** [comment]
Edit: There was a bunch of confusing code here that muddied the waters more than helping anything, removed because I don't think it's useful. Didn't delete because @myz-dev has a useful reply that I don't want to orphan

**Reply by myz-dev** [reply]
Hi,
I do not know if I understand what your goal is. I assume you want to achieve something like this: 
Have a responder that does not only return different variations of an error type, but also, on success, return the type that the client expects on a successful request.

So what I have in my code base for this scenario:
```
#[derive(Responder, Debug)]
pub enum SomeResponder{
    #[response(status = 200, content_type = "Json")]
    Success(Json<MyPayloadType>),
    #[response(status = 400, content_type = "plain")]
    Failure(String),
    //there could be more here
}
```
And in my request that has a signature like: `pub async fn get_payload(/**/ -> RollResponder {}` I return these responses like this:
```
// success:
SomeResponder::Success(Json(some_payload))
// failure:
return SomeResponder::Failure(format!("ERROR!"));
```
Does this help you at all?
